### ISSUE TITLE
[Core][State Observability] Dashboard snapshot not returned with GRPC "Deadline Exceeded" error

### ISSUE THREAD

**Issue by galenhwang** [ray-project/ray#27813]
### What happened + What you expected to happen

Periodically getting this error when polling the Ray dashboard for a snapshot. Normally the data field contains the snapshot object.

```python
{
  "result": false,
  "msg": "Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/optional_utils.py", line 95, in _handler_route
    return await handler(bind_info.instance, req)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/snapshot/snapshot_head.py", line 132, in snapshot
    self.get_session_name(),
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/snapshot/snapshot_head.py", line 384, in get_serve_info
    timeout=GCS_RPC_TIMEOUT_SECONDS,
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 149, in wrapper
    return await f(self, *args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 515, in internal_kv_keys
    reply = await self._kv_stub.InternalKVKeys(req, timeout=timeout)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/aio/_call.py", line 291, in __await__
    self._cython_call._status)
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
  status = StatusCode.DEADLINE_EXCEEDED
  details = "Deadline Exceeded"
  debug_error_string = "{"created":"@1660254637.037033722","description":"Error received from peer ipv4:172.31.72.145:57379","file":"src/core/lib/surface/call.cc","file_line":1074,"grpc_message":"Deadline Exceeded","grpc_status":4}"
>
",
  "data": {}
}
```

Cluster: https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_CHYmiZfmriQ7rBhJCGNY2Bae
Endpoint is `/api/snapshot`.

### Versions / Dependencies

releases/2.0.0rc1

### Reproduction script

check source of test which is creating the cluster

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by jjyao** [comment]
cc @rkooo567 @rickyyx

**Comment by jaikumarg1** [comment]
@alanwguo who is the right owner for this? Assuming core team here.

guessing we can easily reproduce this, if yes, let's write a repro case and assign to Chen

**Comment by alanwguo** [comment]
Yes,

This is happening in the `long_running_actor_deaths` nightly test. We can ping the core oncall next time this happens.

@scv119 does that work?

CC: @galenhwang

**Comment by shawnpanda** [comment]
I am noticing the same failure for the `long_running_many_ppo_1660801565` test. One thing that stood out for me is that in my head node cpu usage monitoring, ray_container gets up to 55% of the CPU while promtail_container's cpu trends up over time up to 40% so that the entire head node cpu is being used up. 

<img width="592" alt="Screen Shot 2022-08-18 at 6 53 36 PM" src="https://user-images.githubusercontent.com/9221501/185525499-f2c92a25-51d6-4473-a3bb-fbd94550325f.png">

**Comment by jaikumarg1** [comment]
@jjyao @rkooo567 any update on this?

**Comment by rkooo567** [comment]
So based on the error output, it is clear the dashboard replies. The failure happens from dashboard -> GCS path. I found out the current timeout is only 3 seconds which is really low (we already observed GCS can have 5+ seconds overhead when it is busy from other releases testst. Note that fixing this is in our radar). I will increase this to match the prod side client timeout (30 seconds)

Normally there are 3 cases where GCS response can be slow

1. Too big responses. This means the seriazliation & deserialization cost is big and saturate the network cost. We are limiting the response size not, so it is not a problem 
2. GCS is overloaded. It can happen when there are e.g., lots of nodes or actors being scheduled. It is a known issue, and we are planning to handle it in from the core side (cc @iycheng )
3. GCS has not enough CPU to handle API calls. -> Seems like this is the case from the above graph (promtail uses too much CPU). This should be handled by the head node resource isolation project.
